### DISCUSSION TITLE
Content Type, "Bag" Parts: how to limit the number of items a user can create in a "Bag" Parts?

### DISCUSSION THREAD

**Discussion by MarGraz** [OrchardCMS/OrchardCore#15549] [Q&A]
Hi everyone,

I've created a `Content Type` named `User`, which I intend to use in an Orchard Core headless microservice. This service will enrich the user information provided by the SSO microservice. The `User` content type should be similar to the following JSON structure:

```json
{
  "SsoUserId": "ajaolkaolk2",
  "Information": {
    "ContactPerson": "Name",
    "Email": "email@example.com",
    "Phone": "123456789",
    "Street": "Pippo Blvd",
    "StreetNumber": "20",
    "City": "Pluto",
    "ZipCode": "12345",
    "Logo": "https://www.example.com/logo.jpg"
  }
}
```

Additionally, I've created an `Information` content type. In the `User` content type, I've added a `bag` part, as shown here:

![image](https://github.com/OrchardCMS/OrchardCore/assets/19241483/4d4ba378-2603-408d-9514-1a6fa010b9a0)

Within the `bag` settings, I've chosen to include the `Information` content type:

![image](https://github.com/OrchardCMS/OrchardCore/assets/19241483/eb6b8c19-a79d-4a3d-adeb-6d41809dcd9b)

**Now from the backend I can create multiple `Information` items, but I only need one.** 

![image](https://github.com/OrchardCMS/OrchardCore/assets/19241483/f968541c-d72c-4d07-baac-d9afe3080ef0)

I used a `bag` **to prevent direct creation of the `Information`** content type and to avoid to use a picker in the `User` content type.

1. Is there a way to restrict the creation to just one `Information` item per user? 
2. Would implementing a `ContentHandlerBase` class be a solution, or is there another method to achieve this from the backend?

Thank you

**Comment by douwinga** [comment]
Should be able to add the settings using a settings display driver for the BagPart `public class BagPartMinMaxSettingsDisplayDriver : ContentTypePartDefinitionDisplayDriver<BagPart>` and add it to the startup `services.AddScoped<IContentTypePartDefinitionDisplayDriver, BagPartMinMaxSettingsDisplayDriver>();`. Then validate using a `ContentPartHandler` or in the `UpdateAsync` method of an additional `ContentPartDisplayDriver<BagPart>`. You could also override the edit view for bag parts to add some javascript to validate on the client browser too.

We did a feature like this at work and it works well, but our needs are different from most peoples. Definitely doable though.

**Reply by MarGraz** [reply]
@douwinga, thank you for your reply.

By using `ContentTypePartDefinitionDisplayDriver<BagPart>`, I will end up displaying the `max` and `min` setting fields when creating a new `User` item on the _"Content" > "Content Items"_ `User` page, correct?

**If yes, I would prefer not to show these settings there**. Instead, **I prefer to include the `min` and `max` settings on the `Bag` edit page** (as we discussed [here](https://github.com/OrchardCMS/OrchardCore/discussions/15522#discussioncomment-8832770)), and then performing the validation when a new `User` is submitted.

To clarify, my intention is to add the `max` and `min` fields to the `Bag` edit page, as shown here:

![image](https://github.com/OrchardCMS/OrchardCore/assets/19241483/eb6b8c19-a79d-4a3d-adeb-6d41809dcd9b)

I plan to proceed with implementation **if this approach is correct**.

Thank you

**Reply by MarGraz** [reply]
@douwinga in the previous comment I confused the `ContentTypePartDefinitionDisplayDriver<BagPart>` you suggested, with the `ContentPartDisplayDriver<MyPart>`. 

So using `ContentTypePartDefinitionDisplayDriver<BagPart>` I should be able to add the `max` and `min` fields to the `Bag` edit page.
I'll start working on implementing this.

Thank you

**Reply by douwinga** [reply]
The `ContentTypePartDefinitionDisplayDriver<BagPart>` -> `Edit` / `EditAsync` methods will build a shape that will show on the bag part settings page.

**Reply by MarGraz** [reply]
@douwinga it works, for the `Edit` method override I took inspiration from the `BagPartSettingsDisplayDriver` class [here](https://github.com/OrchardCMS/OrchardCore/blob/e3176958d137b839e09d49bdb7ca1e298b68ed39/src/OrchardCore.Modules/OrchardCore.Flows/Settings/BagPartSettingsDisplayDriver.cs#L28), and it works, as you can see with a placeholder text:

![image](https://github.com/OrchardCMS/OrchardCore/assets/19241483/36f6fc00-abdf-41d8-9c05-56bdea02ca71)

**How can I know all the available locations to move my `CustomBagPartSettings.Edit.cshtml` from the bottom of the page?**

For what I see, in the `ContentTypeDefinition` > `Parts` > `Settings` property, there's an object that contains a `Position` number. **Is this the right way to find the available locations?**
Here an image of what I saw:

![image](https://github.com/OrchardCMS/OrchardCore/assets/19241483/056aa751-23aa-4950-8cc6-e08a74c134d9)

Find the positions can be very tricky. **Could be nice to have a feature for both frontend and backend, that when turned on, shows the available positions of each shape**. Other CMS like Joomla has this feature, [here](https://docs.joomla.org/Module_Position/en#Module_Map) an example of "Module Map":

![image](https://github.com/OrchardCMS/OrchardCore/assets/19241483/4db8e7e2-8364-4872-a95b-1015117d25d8)

**About the placement topic** I also opened [this discussion](https://github.com/OrchardCMS/OrchardCore/discussions/15464) some days ago.

Thank you

**Reply by douwinga** [reply]
I do not know the best way to find where things can be placed. It depends on the theme, so I guess you would have to look there if the theme doesn't provide a diagram like you have in your screenshot. Maybe `Design` -> `Widgets` in the admin gives a clue for the built in themes, but I honestly just don't know how people are expected to figure that out without just looking at the theme view files.

As for the position, you can include that with your location like `.Location("Content:6")` where `6` is the position number. So if there are items in positions 1-5 then this item will be in the 6th position.
